velox/exec/HashProbe.cpp

```diff
@@ -1056,7 +1059,13 @@ bool HashProbe::maybeReadSpillOutput() {
 RowVectorPtr HashProbe::createFilterInput(vector_size_t size) {
   std::vector<VectorPtr> filterColumns(filterInputType_->size());
   for (auto projection : filterInputProjections_) {
-    ensureLoadedIfNotAtEnd(projection.inputChannel);
+    // Right semi joins only use the probe side for filtering.
+    if (isRightSemiFilterJoin(joinType_) || isRightSemiProjectJoin(joinType_)) {
```

Comment:
If the columns are used in filter only, we should be able to keep them lazy before evaluation

Reply:
I'm not sure that's safe in all cases, which is why I kept this in to maintain the same behavior for right semi joins.

I'm fairly certain they're not impacted by this bug, just logically and through testing.

The reason for ensureLoadedIfNotAtEnd is if we the number of output rows for the given input batch size is larger than max output batch size, we will process the filter for the first part of the batch, and then the next part of the batch on the next iteration, and so on.

For right semi joins we don't output until all of the probe side input has been processed (so we know which rows are filtered and not) so that max output batch size is not relevant, but looking at the code it looks like it's still applied.

I may be missing something in the code or misreading it, but given the complexity of this code and the fact the testing for it is clearly not complete, I'd rather leave fixing this as a future optimization rather than risk introducing another correctness bug in this change.

Reply:
But the source of the bug coming from we read the result that is skipped during lazy eval right?  If we are not using the value outside the filter expression, there is no second use of the same vector thus there will be no bug.

Reply:
This is essentially the same treatment as multi-referenced fields in `FilterProject`: https://github.com/facebookincubator/velox/blob/a6bb4d81fbf52d90081370addafac2b0498ab077/velox/exec/FilterProject.cpp#L138-L140

Reply:
That's right, the bug I'm fixing here does not apply to right semi joins. 

The existing code already calls ensureLoadedIfNotAtEnd for right semi joins, I'm worried that by removing it we would introduce a new bug.  The filter may be evaluated multiple times on the same (Lazy)Vector, just different ranges of it, if the number of hits exceeds outputBatchSize_.

If we are in this case where the filter may be evaluated on a different range of the same Vector later, results_.atEnd() is false by the time we get to evalFilter (this iterates over the hits from a given input batch).  ensureLoadedIfNotAtEnd only loads LazyVectors if results_.atEnd() is false.  So even if I'm wrong, worst case when we call ensureLoadedIfNotAtEnd results_.atEnd() will always be true and the function will return immediately (no LazyVectors will be loaded).

Reply:
Chatted offline: I had misunderstood, we only need to call ensureLoaded if the column is projected out, otherwise we only need to call ensureLoadedIfNotAtEnd, I'll update the PR